Issue #13345: Enabled NoCodeInFileCheckExamplesTest

MANISH-K-07 opened the pull request:
Part of #13345 

Enabled NoCodeInFileCheckExamplesTest

Based on findings in this PR #14395  and Response to  https://github.com/checkstyle/checkstyle/pull/14395#discussion_r1483020974 ,
Updated the `NoCodeInFileCheck` description in all related files.

**_Previously :_** 

`Checks whether file contains code. Files which are considered to have no code:`

**_After update :_** 
```
 * Checks whether file contains code.
 * Java compiler is not raising errors on files with no code or all commented out.
 * Files which are considered to have no code:
```

MANISH-K-07 commented:
All the failing checks are due to the following error only :


[INFO] [checkstyle] Running Checkstyle  on 218 files
[ERROR] [checkstyle] [ERROR] /home/circleci/project/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/nocodeinfile/Example1.java:1: Example test input must declare exactly one class matching 'Example\d+' [exampleClassName]
[ERROR] [checkstyle] [ERROR] /home/circleci/project/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/nocodeinfile/Example2.java:1: Example test input must declare exactly one class matching 'Example\d+' [exampleClassName]
[INFO] ------------------------------------------------------------------------


However, the check input files are needed to be empty !!
Should I add a suppression or is there any alternative ?....

Please suggest on how to proceed.

romani commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/nocodeinfile/Example1.java:
```diff
@@ -6,6 +6,8 @@
 // xdoc section -- start
-// single-line comment // violation
+// single-line comment
 // xdoc section -- end
```
```
Example1 {
// single-line comment
}
```

MANISH-K-07 replied:
Done 👍🏻

romani reviewed (changes requested):
Items

romani commented:
Please fix CI

romani commented:
GitHub, generate website

romani commented:
Sorry I misunderstood a Check functionality.
It should complain on file that has no java code.

https://github.com/checkstyle/checkstyle/blob/8d5daa51a7e9e9b294f549eaace22614b944bea6/src/main/java/com/puppycrawl/tools/checkstyle/checks/NoCodeInFileCheck.java#L84-L86

So in this case we need to move such java files to non-compileable folder, keep comments with Check config, remove package declaration, remove class declaration. Keep comment. Both files should have violation on first line of file.

Under comment with config, put violation comment that points to first line.

Please make first file to have whole class completed out:
```
// public class Example1 {
// // Whole code was commented out , violation on first line of file 
// }
```

Second file:
```
/*
public class Example2 {
 // Whole code was commented out , violation on first line of file 
}
*/
```

romani reviewed (changes requested):
Comment is above.

MANISH-K-07 commented:
> Sorry I misunderstood a Check functionality.
> It should complain on file that has no java code.

@romani ,

Exactly !! 
I was wondering why you would suggest to add a class in an input file of `NoCodeInFileCheck`. That would be defeating the sole purpose of the check as even a single line is considered code and there is no violation thrown.

Anyways....
I have modified the files now. Moved them to resources-noncompilable and enabled the check to retrieve path.
**_Please review_**

```
INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS                                                                                                                  
[INFO] ------------------------------------------------------------------------                                                       
[INFO] Total time:  06:21 min                                                                                                         
[INFO] Finished at: 2024-02-02T13:26:01+05:30                                                                                         
[INFO] ------------------------------------------------------------------------                                              
```

MANISH-K-07 commented:
The failing checks are irrelevant to my PR ??

romani commented:
GitHub, generate website

romani reviewed (changes requested):
Items

MANISH-K-07 commented on src/xdocs-examples/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/nocodeinfile/Example1.java:
```diff
@@ -0,0 +1,16 @@
+  </module>
+</module>
+*/
+//non-compiled with javac: Compilable with Java14
```
That wasn't intentional.. must have missed it during copy-paste
Sorry about that. Removed now

MANISH-K-07 commented:
@romani 

Getting the following error while trying to edit the `non-compiled` comment : 
```
[INFO] [checkstyle] Running Checkstyle  on 218 files
[ERROR] [checkstyle] [ERROR] /home/circleci/project/src/xdocs-examples/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/nocodeinfile/Example1.java:1: Missing compilation comment [presenceOfCompilationComment]
[ERROR] [checkstyle] [ERROR] /home/circleci/project/src/xdocs-examples/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/nocodeinfile/Example2.java:1: Missing compilation comment [presenceOfCompilationComment]
```

The checks are not allowing any `non-compiled` comment other than the format from file [checkstyle-examples-checks.xml](https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle-examples-checks.xml) : 
```
    <property name="id" value="presenceOfCompilationComment"/>
    <property name="format" value="^//non-compiled with javac: Compilable with Java(\d+)"/>
    <property name="message"
              value="Missing compilation comment"/>
```

It is obvious from this PR #14395 that `Compilable with Java(\d+)` is not always the case.

We could change the pattern to check something like ` //non-compiled with javac: + "some related message"`

Or is my understanding wrong somewhere ?....
How would you suggest we proceed further with this ?

romani commented:
Before we start hacking, can we put it back to compiled folder?

MANISH-K-07 commented:
There is still no code in the files

> Before we start hacking, can we put it back to compiled folder?

Won't that take us back to this error ?
```
[INFO] [checkstyle] Running Checkstyle  on 218 files
[ERROR] [checkstyle] [ERROR] /home/circleci/project/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/nocodeinfile/Example1.java:1: Example test input must declare exactly one class matching 'Example\d+' [exampleClassName]
[ERROR] [checkstyle] [ERROR] /home/circleci/project/src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/nocodeinfile/Example2.java:1: Example test input must declare exactly one class matching 'Example\d+' [exampleClassName]
[INFO] ------------------------------------------------------------------------

```

romani commented:
please suppress such violation like 
https://github.com/checkstyle/checkstyle/blob/c3104f1b4dfade09a0afeb064a58cbf76cfd6a42/config/checkstyle-examples-suppressions.xml#L9-L11

with comment `no class in java file by design`

MANISH-K-07 commented:
From  https://github.com/checkstyle/checkstyle/pull/14395#issuecomment-1919212488 ,

> However, the check input files are needed to be empty !!
> Should I add a suppression or is there any alternative ?....

We eventually got back to square one  :)

> please suppress such violation like
> with comment `no class in java file by design`

Done now..


**Regarding the failing CI,**
I am not entirely sure why [ci/semaphoreci/pr: Checkstyle CI pipeline on Semaphore](https://checkstyle.semaphoreci.com/workflows/245f7dbf-6557-4041-a2fb-403870b2511e?pipeline_id=3c673e96-954b-464e-8bdc-72841d68d29c)  keeps timing out. The build isn't exactly a failure as there are no errors, it is just incomplete. Observed the same while reviewing another PR ( https://github.com/checkstyle/checkstyle/pull/14439#pullrequestreview-1866665996 )

MANISH-K-07 commented:
Github, generate website

romani commented:
```
rivanov@p5510:/var/tmp$ cat Test.java
/*
SummaryJavadoc
violateExecutionOnNonTightHtml = (default)false
forbiddenSummaryFragments = (default)^$
period = (default).


*/

// package com.puppycrawl.tools.checkstyle.checks.javadoc.abstractjavadoc;
// // violation 2 lines below 'Summary of Javadoc is missing an ending period.'
// // violation below 'First sentence of Javadoc is missing an ending period.'
// /**Javadoc*/
// class InputAbstractJavadocCache1{}
rivanov@p5510:/var/tmp$ javac Test.java
rivanov@p5510:/var/tmp$ 

```


empty files are compile-able so we need to put them back to compiled folder and suppress all Checkstyle validations that do not allow file to be empty.

romani commented on config/checkstyle-examples-suppressions.xml:
```diff
@@ -6,6 +6,12 @@
 
 <suppressions>
 
+  <!-- These files have no class in java file by design. -->
```
not related to this line.

Please update description of Check from `Checks whether file contains code. Files which are considered to have no code: ` to `Checks whether file contains code. Java compiler is not raising errors on java files that has no code or all is commentend out. Files which are considered to have no code: `

MANISH-K-07 replied:
I have updated `NoCodeInFileCheck.java`
Should I apply the same for [InputJavadocMetadataScraperNoCodeInFileCheck.java](https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/meta/javadocmetadatascraper/InputJavadocMetadataScraperNoCodeInFileCheck.java) also ?

romani reviewed (changes requested):
items:

MANISH-K-07 commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/nocodeinfile/Example2.java:
```diff
@@ -6,8 +6,10 @@
+
 // xdoc section -- start
-/* // violation
+/* // violation, block comment is not code
```
So, after restoring commented out classes and `// violation` pointing to first line  gives no errors. If we put commented out class, there is no need of `exampleClassName suppressions` that I have previously added.

However, putting the violation comment _on_ first line of file gives the following expected error as we are meddling with the config comment : 

```
[ERROR] Errors:                                                                                                                       
[ERROR]   NoCodeInFileCheckExamplesTest.testExample1:40->AbstractModuleTestSupport.verifyWithInlineConfigParser:235 » Checkstyle Confi
g comment not specified properly in E:\MANISH\Workspace\checkstyle\src\xdocs-examples\resources\com\puppycrawl\tools\checkstyle\checks
\nocodeinfile\Example1.java
[ERROR]   NoCodeInFileCheckExamplesTest.testExample2:49->AbstractModuleTestSupport.verifyWithInlineConfigParser:235 » Checkstyle Confi
g comment not specified properly in E:\MANISH\Workspace\checkstyle\src\xdocs-examples\resources\com\puppycrawl\tools\checkstyle\checks
\nocodeinfile\Example2.java

```

romani replied:
>  If we put commented out class, there is no need of exampleClassName suppressions that I have previously added.

even better for us, the less suppression the better.

romani replied:
> However, putting the violation comment on first line of file gives

you can place violation comment below comment with config and make it `// violation X lines above`

MANISH-K-07 replied:
> > However, putting the violation comment on first line of file gives
> 
> you can place violation comment below comment with config and make it `// violation X lines above`

Yepp, the commit currently holds the same....

romani commented:
@MANISH-K-07 , please fix CI
https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=19572&view=logs&j=c902ebb4-c9f8-5f09-4e17-ff78fbbc842e&t=9ca98c81-ff64-58f0-9d03-a23ac1c4a111&l=333
a bit more changes are required for Check description changes.

romani commented on src/xdocs/checks/misc/nocodeinfile.xml:
```diff
@@ -45,12 +45,16 @@
           Content of the files:
         </p>
         <source>
-// single-line comment // violation
```
Please remove line `Content of the files:` that is 3 lines above this line.
In final rendering it looks weird

MANISH-K-07 replied:
Removed....

romani commented on src/main/resources/com/puppycrawl/tools/checkstyle/meta/checks/NoCodeInFileCheck.xml:
```diff
@@ -5,7 +5,9 @@
              parent="com.puppycrawl.tools.checkstyle.TreeWalker">
          <description>&lt;p&gt;
- Checks whether file contains code. Files which are considered to have no code:
+ Checks whether file contains code.
```
You need to update template file that by this change

MANISH-K-07 replied:
Done, Updated the Check description too !!

romani reviewed (changes requested):
Items

MANISH-K-07 commented:
GitHub, generate website

MANISH-K-07 commented:
@romani ,
This is off-topic but could you please review PR #14397 . It's been sitting idle for almost two weeks now 
 :)

romani commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/nocodeinfile/Example2.java:
```diff
@@ -6,8 +6,12 @@
 // xdoc section -- start
-/* // violation
- block comment
+/*
```
Above this line and below `// xdoc section` please add comment `// the violation on first line of file `
 Do same for another example.

Reason: final Rendering of web page looks strange and not clear where is violation

MANISH-K-07 replied:
> final Rendering of web page looks strange and not clear where is violation

I agree, added the message

romani reviewed (changes requested):
Last

MANISH-K-07 commented:
Github, generate website

romani reviewed (approved):
Ok to merge.

Thanks a lot, you gave this Check documentation a new life

MANISH-K-07 commented:
> Ok to merge.
> 
> Thanks a lot, you gave this Check documentation a new life

Always Happy to help !!
`:)`
